OverloadingOperations.cs

De-Len opened the pull request.

De-Len committed:
OverloadingOperations.cs

De-Len committed:
Update Program.cs

Добавил пользовательские исключения. Подправил код.

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,535 @@
+
+    public int GetLength { get; set; }
+    public double Determinant { get; set; }
+    static readonly Random random = new Random();
```
Поля\свойства либо с большой буквы, либо с маленькой. Тут надо random с большой буквы

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,535 @@
+    public SquareMatrix(int size)
+    {
+        Matrix = new int[size, size];
+        for (var rowOfMatrix = 0; rowOfMatrix < Matrix.GetLength(0); ++rowOfMatrix)
```
здесь то же самое - _R_owOfMatrix

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,535 @@
+            throw new SquareMatrixDimensionsException("Матрицы разного размера.", Rows, matrix2.Value.GetLength(0));
+        }
+
+        int[,] result = new int[Rows, Columns];
```
result не слишком понятная переменная

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,535 @@
+    int[,] Matrix;
+    public int[,] Value { get { return Matrix; } }
+
+    public int GetLength { get; set; }
```
Get - слово, присущее для метода. Для поля нужно другое. Например Length
Также не вижу, чтобы данное поле что-то возвращало, геттер не прописан

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,535 @@
+
+    //ПРОЧИЕ МЕТОДЫ
+
+    public static int[,] ConvertMatrixToDouble(double[,] inputMatrix) // Приведение типов
```
Метод вроде должен возвращать матрицу с типом дабл, но при этом возвращает матрицу с типом инт

De-Len replied:
Это же приведение типов? Насколько я понял, это изменение типа данных?

Hamjoshua replied:
Рекомендую к прочтению эту статью

https://metanit.com/sharp/tutorial/3.37.php

De-Len committed:
Update Program.cs

Исправил инкременты и дефект обратной матрицы 2x2.

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,535 @@
+    }
+
+}
+class SquareMatrixDimensionsException : Exception
```
Вижу кастомную ошибку, хорошая работа

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,535 @@
+        }
+
+        Console.WriteLine("Прототип первой матрицы:");
+        ExampleMatrix1.Clone();
```
В чем смысл этой строки? Значение просто возвращается и ни к какой переменной не стыкуется. Можно было бы ее запихать в вывод, но у тебя уже есть метод для вывода объекта PrintClone. Следовательно, лучше убрать эту строку

Hamjoshua reviewed (commented):
Небольшие дефекты в наименовании и логики

De-Len committed:
Update Program.cs

Исправлено по ТЗ тимлида.

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+
+    public double[,] InvertMatrix(int[,] matrix)
+    {
+        int LenthgtOfMatrix = matrix.GetLength(0);
```
*LengthOfMatrix

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+                subMatrix[Row, ++Column] = matrix[IndexOfRow, IndexOfColumn];
+            }
+        }
+        return subMatrix;
```
subMatrix с маленькой буквы

лучше сразу условиться - переменные с большой или маленькой? В соглашении написано, что PascalCase. Но половина интернета безбожно пишет с маленькой. Этих грешников можно простить, их большинство

Если решил следовать силам света, то следи, чтобы все переменные (кроме полей аргументов и приватных переменных, которые еще снабжаются "_") были с большой буквы

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+        return TransposedMatrix;
+    }
+
+    public void PrintMatrix()
```
Я считаю все Print методы избыточными. Задача этого класса - давать результат, дискриминант, обратную матрицу и т.д. О выводе информации в читаемом варианте пусть заботится другой модуль

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+        }
+        catch (Exception)
+        {
+            Console.WriteLine($"Ошибка: Вы ничего не ввели!");
```
А что, если пользователь в поле для ввода размера введет не пустую строку, а просто строку? Например "аоыафа". Пользователь будет удивлен, когда на "аоыафа" программа ему скажет, мол он ничего не ввел. Есть два стула - либо сделать общий вывод "Неправильный ввод!", либо же обработать каждую ошибку
В случае поступления пустой строки - вы ничего не ввели
В случае поступления букв и прочего, что не является цифрами - вы ввели не цифры

Hamjoshua replied:
К тому же, программа не останавливается и не пытается получать эти значения заново. Она просто продолжается, что может привести к ошибке кода. Лучше зациклить ввод значений, пока не пользователь не введет значение в правильном формате

Hamjoshua replied:
Кстати, пользователь может ввести -1. Это тоже нужно обработать

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+        }
+    }
+
+
```
Следи за тем, чтобы у тебя не было слишком много пустого пространства

Но и за тем, чтобы оно было: Между логическими блоками должен быть один Enter

Hamjoshua replied:
Между функциями Enter нужен в любом случае. Как и между классами. Enter'ы внутри кода по твоему усмотрению, но советую членить логически связанные блоки Enter'ами

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+
+    public int Length { get; set; }
+    public double Determinant { get; set; }
+    static readonly Random Random = new Random();
```
"При работе с полями static, являющимися private или internal, используйте префикс s_, а для статического потока используйте t_."

Здесь: s_camelCase

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+
+class SquareMatrix
+{
+    int[,] Matrix;
```
"При именовании private или internal полях и префиксе _их используйте верблюдьи регистры ("верблюдьи"). Используйте регистр верблюда при именовании локальных переменных, включая экземпляры типа делегата."

Приватные поля и свойства нужно правильно именовать - _camelCase

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+
+    private int[,] GetAdjointMatrix(int[,] matrix)
+    {
+        int LenghtOfMatrix = matrix.GetLength(0);
```
*LengthOfMatrix. Перепроверь правильное написание слов

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+class SquareMatrix
+{
+    int[,] Matrix;
+    public int[,] Value { get { return Matrix; } }
```
Value применимо, если в нем содержится одно значение. Если значений несколько, лучше назвать поле по-другому. Items, к примеру

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+        return !(matrix1 == matrix2);
+    }
+
+    //ПРОЧИЕ МЕТОДЫ
```
между "//" и комментарием должен быть пробел
"// хороший комментарий"

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,545 @@
+
+
+
+class Program
```
Пока задачи небольшие, можно размещать в одном файле несколько объектов - исключения, классы, матричный калькулятор
Но задачи будут усложняться, классы раздуваться, строки доходить до 1000-ных значений. Это будет тяжко читать
На будущее - разделяй объекты на отдельные файлы. В одном пусть лежат все ошибки, в другом все возможные варианты матрицы, в третьем матричный калькулятор

Hamjoshua reviewed (changes requested):
Последний рывок и программа будет готова к мерджу на test

De-Len committed:
Update Program.cs

Corrected type conversion
Added exceptions
Fixed naming of variables

De-Len committed:
Update Program.cs

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,569 @@
+{
+    int[,] _matrix;
+    public int[,] Value { get { return _matrix; } }
+    public int Length { get; set; }
```
Это поле не содержит никакой логики в геттере/сеттере, но при этом используется 4 раза. Что оно должно возвращать, если оно никак не задается?

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,569 @@
+    int[,] _matrix;
+    public int[,] Value { get { return _matrix; } }
+    public int Length { get; set; }
+    public double Determinant { get; set; }
```
Поле было объявлено, но ни разу не использовалось

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,569 @@
+        {
+            throw new SquareMatrixDimensionsException("Матрица вырожденная, обратной матрицы не существует.");
+        }
+        double Determinant = CalculateDeterminant(matrix);
```
у тебя CalculateDeterminant вызывается два раза. Это значит, что машина два раза пройдется по сложному алгоритму в случае ненулевого детерминанта

Здесь лучше заранее найти детерминант, использовать его в условии и работать с ним дальше

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,569 @@
+                Console.WriteLine("Введите размер второй квадратной матрицы (одно число):");
+                InputSizeOfMatrix2 = (Console.ReadLine());
+                SizeOfMatrix2 = int.Parse(InputSizeOfMatrix2);
+                if (SizeOfMatrix1 != 0 &&  SizeOfMatrix2 != 0)
```
Число может быть отрицательным, и здесь лучше написать условие > 0

Hamjoshua replied:
Также пользователю будет не совсем ясно, почему он ввел 0, а его начали снова просить ввести размеры матриц. Пользователя сначала нужно "отругать" за неправильное действие, а потом просить ввести размеры заново

сделай блок else, который будет подсказывать "Размер должен быть больше 0!"

Hamjoshua commented on Program.cs:
```diff
@@ -0,0 +1,569 @@
+
+    public static SquareMatrix operator *(SquareMatrix matrix1, SquareMatrix matrix2)
+    {
+        int Rows1 = matrix1.Value.GetLength(0);
```
Не самое лучшее имя для переменной, что означает количество рядов. Потому что под Rows вполне воспринимается массив/список/коллекция из рядов матрицы. Лучше назвать RowCount, или RowLength. И так со всеми. 

Цифру в конце писать можно

Hamjoshua replied:
Поменяй везде, ибо Row/Column в данном контексте воспринимается как одномерный массив. У тебя же это имя переменной применятся в случаях, где участвуют индексы этих массивов. Повнимательнее

Hamjoshua reviewed (changes requested):
Требуются небольшие изменения

De-Len committed:
Update Program.cs

Added exception handler
deleted variable "Length" from "SquareMatrix"
Fixed "Calculate Detetminant" - going out once time (Можно ли на сленге так сказать? - один раз теперь используется)
Fixed naming of variables in overloading of *

De-Len committed:
Update Program.cs

Deleted unused variable

Hamjoshua reviewed (approved):
Все четко
